run_bin.go

```diff
@@ -85,7 +85,7 @@ func (c *CoverageCollector) TearDown() error {
 }
 
 // RunBinary runs the instrumented binary at binPath with env environment variables, executing only the test with mainTestName with the specified args.
-func (c *CoverageCollector) RunBinary(binPath string, mainTestName string, env []string, args []string) (output string, exitCode int, err error) {
+func (c *CoverageCollector) RunBinary(binPath string, mainTestName string, env []string, args []string, stdinInput string) (output string, exitCode int, err error) {
```

Comment:
Yeah I agree, you want to make this more generic probably.

This change is akin to the "telescoping constructor" pattern (which doesn't work in golang since you don't have method overloading, hence a breaking change). The typical response to this is to refactor the interface to use a Builder object pattern instead.

You could probably make this change in a backward compatible manner if you just have this existing method delegate to a Builder object under the hood.

Reply:
@codyaray thanks for the feedback! My initial thought is to go ahead with a callback function since this seems to expose the cmd object to the user nicely, so that the StdInPipe() function and other cmd object can be accessed. Do you see any tradeoffs between doing this versus the builder pattern? 

My thought is that the builder pattern would require a new builder for every action people might want to perform on cmd object in the future? But I guess adding the callback parameter makes things not backward-compatible? Unless it's a variadic parameter like @mpokryva posted above ... is it a bad pattern to have a variadic parameter just for the sake of backward compatibility? I guess there could be a case where a user has modularized all their potential callback actions, and would pass a series of them to RunBinary as required, so maybe the variadic callback parameter makes sense...

Reply:
sorry that's a lot of thoughts 😆  lmk there's anything I can clear up

Reply:
Could do things in the style of [go-kit](https://github.com/go-kit/kit/blob/e2d71a06a40aa95cb82ccd72e854893612c02db7/transport/httprp/server.go#L40)
```
type CmdFunc func(cmd *exec.Cmd) error
type CoverageCollectorOption func(collector *CoverageCollector)

func PreExec(cmdFuncs ...CmdFunc) CoverageCollectorOption {
	return func(c *CoverageCollector) {
		c.preCmdFuncs = cmdFuncs
	}
}

func PostExec(cmdFuncs ...CmdFunc) CoverageCollectorOption {
	return func(c *CoverageCollector) {
		c.postCmdFuncs = cmdFuncs
	}
}

// RunBinary runs the instrumented binary at binPath with env environment variables, executing only the test with mainTestName with the specified args.
func (c *CoverageCollector) RunBinary(binPath string, mainTestName string, env []string, args []string, options ...CoverageCollectorOption) (output string, exitCode int, err error) {
	for _, option := range options {
		option(c)
	}
```

Reply:
I've seen variadic params for the sake of backwards compatibility, and yeah it's pretty inelegant. The above solution lends itself naturally to variadic params though, so I don't think that's bad

Reply:
ooh nice @mpokryva I'll check out doing it like that ... looks pretty slick 😎  🆒